### PR TITLE
chore(compas-indexes): store real indexes and in-progress indexes as separate lists

### PR DESCRIPTION
**Pull request by lerouxb**
Extracted out of https://github.com/mongodb-js/compass/pull/6290

The basic idea is to not merge real indexes and in-progress indexes in the store. That's error prone because you can't really merge them together. All we have to go on is index name and if you override either the in progress details or the real index details it leads to bugs either way. We poll the real details now anyway, so it is eventually consistent as it is. With rolling indexes this just gets worse.

Furthermore, the types aren't really very compatible. We have an index name and little more for in-progress indexes at the moment and for rolling indexes we'll have a name and a type. If we want type and properties for a regular in-progress index we have to infer that based on the input ourselves and in future if we want properties for rolling indexes we have to map those from atlas' response type ourselves. Similar for any other field we might display in the table. Things like size and usage count can only be zero in those cases too. So I decided to have three types (two in this PR) and corresponding lists, then we concatenate them together in the table.

This happens to fit redux's recommendation to not store derived state in the store.

The rest of the changes are caused by that basic decision or just drive-by fixes of bugs I spotted while working on it. Partially caught by having narrower types for regular and in progress indexes.

### CODE REVIEW

File: packages/compass-indexes/src/components/regular-indexes-table/index-actions.tsx [hunk 1]
```diff
@@ -32,16 +51,13 @@ const IndexActions: React.FunctionComponent<IndexActionsProps> = ({
...
-        label: `Drop Index ${index.name}`,
-        icon: 'Trash',
-      },
-    ];
+    const actions: GroupedItemAction<IndexAction>[] = [];
 
-    if (serverSupportsHideIndex(serverVersion)) {
-      actions.unshift(
+    if (
+      index.compassIndexType === 'regular-index' &&
```
Comment (lerouxb) [inline packages/compass-indexes/src/components/regular-indexes-table/index-actions.tsx:57]: This was a bug - we showed the hide button for in-progress indexes.

File: packages/compass-indexes/src/components/regular-indexes-table/index-actions.tsx [hunk 2]
```diff
@@ -58,6 +74,19 @@ const IndexActions: React.FunctionComponent<IndexActionsProps> = ({
       );
     }
 
+    // you can only drop regular indexes or failed inprogress indexes
```
Comment (lerouxb) [inline packages/compass-indexes/src/components/regular-indexes-table/index-actions.tsx:77]: This was also a bug - you could drop an in progress index that was really still in progress and that just wouldn't work because the index creation isn't failed and it isn't a real index yet either.

File: packages/compass-indexes/src/components/indexes/indexes.spec.tsx [hunk 3]
```diff
@@ -95,7 +95,9 @@ describe('Indexes Component', function () {
       },
     });
     expect(screen.getByTestId('indexes-toolbar')).to.exist;
-    // TODO: actually check for the error
+    expect(screen.getByTestId('indexes-error').textContent).to.equal(
```
Comment (lerouxb) [inline packages/compass-indexes/src/components/indexes/indexes.spec.tsx:98]: Drive-by: I spotted the TODO when I audited the new TODOs I'm adding in this PR.

File: packages/compass-indexes/src/components/regular-indexes-table/index-actions.tsx [hunk 4]
```diff
@@ -2,10 +2,28 @@ import semver from 'semver';
 import React, { useCallback, useMemo } from 'react';
 import type { GroupedItemAction } from '@mongodb-js/compass-components';
 import { ItemActionGroup } from '@mongodb-js/compass-components';
-import type { RegularIndex } from '../../modules/regular-indexes';
+import type { InProgressIndex } from '../../modules/regular-indexes';
+
+/*
+TODO: we can change this to
```
Comment (lerouxb) [inline packages/compass-indexes/src/components/regular-indexes-table/index-actions.tsx:8]: This one I'm still going to deal with in this PR.

Comment (lerouxb) [reply packages/compass-indexes/src/components/regular-indexes-table/index-actions.tsx]: done.

File: packages/compass-indexes/src/components/regular-indexes-table/regular-indexes-table.tsx [hunk 5]
```diff
@@ -80,12 +121,14 @@ function sortFn(
     return propSort;
   }
 
-  if (field === 'usage') {
-    field = 'usageCount';
-  }
+  const prop = field as unknown as SortableField;
```
Comment (gribnoysup) [inline packages/compass-indexes/src/components/regular-indexes-table/regular-indexes-table.tsx:124]: Nit: if you turn the check at the top of this function into a method that asserts, you won't need to assert types manually here

```ts
function isSupportedSortField(
  field: string
): field is 'name' | 'type' | 'size' | 'usageCount' | 'properties' {
  return ['name', 'type', 'size', 'usageCount', 'properties'].includes(field);
}

// ...

  if (!isSupportedSortField(field)) {
    return 0;
  }
```

File: packages/compass-indexes/src/components/regular-indexes-table/regular-indexes-table.tsx [hunk 6]
```diff
@@ -163,44 +307,40 @@ export const RegularIndexesTable: React.FunctionComponent<
...
-          <IndexKeysBadge
-            keys={index.fields}
-            data-testid={`indexes-details-${index.name}`}
-          />
-        ),
-      })),
+        return {
+          ...indexData,
+          // eslint-disable-next-line react/display-name
+          renderExpandedContent: () => (
```
Comment (gribnoysup) [inline packages/compass-indexes/src/components/regular-indexes-table/regular-indexes-table.tsx:332]: Super tiny nit: if you write this as a named function, you won't need to eslint-disable:

```jsx
          renderExpandedContent() { return (
            // ...
```

**Review by gribnoysup — APPROVED** [review]
Nice cleanup! A few small nits, but they are super optional, great work!

File: packages/compass-indexes/src/components/regular-indexes-table/regular-indexes-table.tsx [hunk 7]
```diff
@@ -67,11 +84,35 @@ function sortByProperties(a: RegularIndex, b: RegularIndex) {
...
+      // TODO(COMPASS-8335): type should be supported by in-progress-index
+      return 'unknown';
+    }
+    if (field === 'size' || field === 'usageCount') {
+      return 0;
+    }
+    return (index as InProgressIndex)[field];
+  }
+
+  return (index as RegularIndex)[field];
```
[the remark this answers is not in the document]
Comment (lerouxb) [reply packages/compass-indexes/src/components/regular-indexes-table/regular-indexes-table.tsx:104]: Aah thanks for pointing this out. I think those were left over from a previous iteration. Maybe with rolling indexes or when I had a more complicated solution. I'll look into it.

File: packages/compass-indexes/src/components/regular-indexes-table/regular-indexes-table.tsx (hunk 6 above)
Comment (lerouxb) [reply packages/compass-indexes/src/components/regular-indexes-table/regular-indexes-table.tsx:332]: Oops. Never even noticed that. I just copy/pasted what we had :) Will fix.
